src/combinatorics/additive/freiman.lean

```diff
@@ -0,0 +1,271 @@
+/-
+Copyright (c) 2021 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+-/
+import data.multiset.basic
+
+/-!
+# Freiman homomorphisms
+
+In this file, we define Freiman homomorphisms. A `n`-Freiman homomorphism is a function `f : α → β`
+such that `f (x₁ * ... * xₙ) = f (y₁ * ... * yₙ)` for all `x₁, ..., xₙ, y₁, ..., yₙ` such that
```

Comment:
isn't it `f(x₁) * f(x₂) * ...` instead?

Reply:
Nope! That would get us back the usual homomorphisms! The rule for a Freiman is that you can never get back up to the domain. There's not many references unfortunately. It seems to be quite a niche concept.

Reply:
I swear `(s.map to_fun).prod` is what I saw, as opposed to `to_fun (s.prod)`, and [Wiki](https://en.wikipedia.org/wiki/Freiman%27s_theorem#Freiman_homomorphisms_and_the_Ruzsa_modeling_lemma) seems to back me up for the additive versions, which are described

Reply:
Oh sorry yeah of course. I thought you meant `f (x_1) * ... * f (x_n) = f (y_1 * ... * y_n)` .